### ISSUE TITLE
Match less general versions of the LHS

### ISSUE THREAD

**Issue by josephcsible** [ndmitchell/hlint#684]
If we have a hint for a more general thing than a piece of code uses, we should fire the hint for the less general thing as if it were for the same thing. For example, we have a hint that recommends `m >>= return` be replaced with `m`. If someone writes `m >>= Just` or `m >>= Right` instead, I'd like that hint to match on it anyway.

**Comment by ndmitchell** [comment]
I'm not sure that's a great idea in general. Perhaps for just `return`/`pure` it might be a good idea? There is already an unhealthy overlap between `pure` and `return`, so whatever fixes that could fix this too.

**Comment by josephcsible** [comment]
Do you have an example of when it wouldn't be a good idea? I can only think of cases when it would be, even for things other than pure/return. For example, I think our rule of `foldr (<|>) empty` to `asum` should also match on `foldr (<|>) Nothing`.

**Comment by ndmitchell** [comment]
I don't have any counterexamples, but I imagine there are some - e.g. given `foldr (++) []` might suggest `mconcat` instead of `concat`. Of course, HLint has no idea what is a more general or specific instance, so someone would need to give it a small fixed list of things that are generalisations.

**Comment by josephcsible** [comment]
Wouldn't it suggest `mconcat` in addition to `concat`, not instead of it? If so, how is that a bad thing?

**Comment by ndmitchell** [comment]
Suggesting mconcat and concat is a bad thing, because concat is the right answer, and better to give one right answer rather than two.

**Comment by josephcsible** [comment]
If we do decide to only do this for some functions and not all of them, we should definitely include `fmap` as one. We have a lot of rules duplicated with both `map` and `fmap` variants, and for those where we don't, useful hints aren't firing when the other one is used.

**Comment by ndmitchell** [comment]
Yeah, my feeling is that some kind of rule that you can have "meta-variables", e.g. `$pure` stands for `pure` or `return`. Then rule authors can choose where they make sense.
